Friend/firmware/firmware_v1.0/src/transport.c

```diff
@@ -89,6 +120,216 @@ struct bt_conn *current_connection = NULL;
 uint16_t current_mtu = 0;
 uint16_t current_package_index = 0;
 
+
+
+
+static uint32_t current_button_time = 0;
+static uint32_t previous_button_time = 0;
+
+const int max_debounce_interval = 700;
+static bool was_pressed = false;
+
+//
+// button
+//
+void button_pressed(const struct device *dev, struct gpio_callback *cb,
+		    uint32_t pins)
+{
+    current_button_time = k_cycle_get_32();
+	if (current_button_time - previous_button_time < max_debounce_interval) { //too low!
+	}
+	else { //right...    
+        int temp = gpio_pin_get_raw(dev,d5_pin_input.pin);
+        if (temp) {
+            was_pressed = true;
+        }
+        else {
+            was_pressed = false;
+        }
+	}
+	previous_button_time = current_button_time;
+}
+#define BUTTON_CHECK_INTERVAL 40 // 0.04 seconds, 25 Hz
+
+void check_button_level(struct k_work *work_item);
+
+K_WORK_DELAYABLE_DEFINE(button_work, check_button_level);
+
+typedef enum {
+    IDLE, 
+    ONE_PRESS,
+    TWO_PRESS,
+    GRACE
+} FSM_STATE_T;
+
+
+#define DEFAULT_STATE 0
+#define SINGLE_TAP 1
+#define DOUBLE_TAP 2
+#define LONG_TAP 3
+#define BUTTON_PRESS 4
+#define BUTTON_RELEASE 5
+
+
+// 4 is button down, 5 is button up
+static FSM_STATE_T current_button_state = IDLE;
+static uint32_t inc_count_1 = 0;
+static uint32_t inc_count_0 = 0;
+
+
+static int final_button_state[2] = {0,0};
+const static int threshold = 10;
+
+
+static void reset_count() {
+    inc_count_0 = 0;
+    inc_count_1 = 0;
+}
+static void notify_press() {
+    final_button_state[0] = BUTTON_PRESS;
+    LOG_INF("pressed");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));
+}
+
+static void notify_unpress() {
+    final_button_state[0] = BUTTON_RELEASE; 
+    LOG_INF("unpressed");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));  
+}
+
+static void notify_tap() {
+      final_button_state[0] = SINGLE_TAP;
+    LOG_INF("tap\n");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));  
+}
+
+static void notify_double_tap() {
+      final_button_state[0] = DOUBLE_TAP; //button press
+    LOG_INF("double tap\n");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));  
+}
+
+static void notify_long_tap() {
+    final_button_state[0] = LONG_TAP; //button press
+    LOG_INF("long tap\n");
+    bt_gatt_notify(current_connection, &button_service.attrs[1], &final_button_state, sizeof(final_button_state));  
+}
+
+#define LONG_PRESS_INTERVAL 50
+#define SINGLE_PRESS_INTERVAL 2
+void check_button_level(struct k_work *work_item) {
```

Comment:
Move logic to a button.c and button.h too much logic here to be inside transport.c